autopilot/betweenness_centrality.go

```diff
@@ -0,0 +1,208 @@
+type stack struct {
+	stack []int
+}
+
+func (s *stack) Push(v int) {
+	s.stack = append(s.stack, v)
+}
+
+func (s *stack) Top() int {
+	return s.stack[len(s.stack)-1]
+}
+
+func (s *stack) Pop() {
+	s.stack = s.stack[:len(s.stack)-1]
+}
+
+func (s *stack) Empty() bool {
+	return len(s.stack) == 0
+}
+
+// queue is a simple int queue to help with readability of Brandes'
+// betweenness centrality implementation below.
+type queue struct {
+	queue []int
+}
+
+func (q *queue) Push(v int) {
+	q.queue = append(q.queue, v)
+}
+
+func (q *queue) Front() int {
+	return q.queue[0]
+}
+
+func (q *queue) Pop() {
+	q.queue = q.queue[1:]
+}
+
+func (q *queue) Empty() bool {
+	return len(q.queue) == 0
+}
+
+// simpleGraph stores a simplifed adj graph of a channel graph to speed
+// up betweenness centrality calculation.
+type simpleGraph struct {
+	// Nodes is a map from node index to NodeID.
+	Nodes []NodeID
+
+	// Adj stores nodes and neighbors in an adjacency list.
+	Adj [][]int
+}
+
+// makeSimpleGraph creates a simplified graph from the current channel graph.
+func makeSimpleGraph(g ChannelGraph) (*simpleGraph, error) {
+	nodes := make(map[NodeID]int)
+	adj := make(map[int][]int)
+	idx := 0
+
+	// Iterate over each node and each channel and update the adj and the node
+	// index.
+	err := g.ForEachNode(func(node Node) error {
+		key := NodeID(node.PubKey())
+		u, ok := nodes[key]
+
+		if !ok {
+			nodes[key] = idx
+			u = idx
+			idx++
+		}
+
+		return node.ForEachChannel(func(edge ChannelEdge) error {
+			key := NodeID(edge.Peer.PubKey())
+			v, ok := nodes[key]
+
+			if !ok {
+				nodes[key] = idx
+				v = idx
+				idx++
+			}
+
+			adj[u] = append(adj[u], v)
+			return nil
+		})
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	graph := &simpleGraph{
+		Nodes: make([]NodeID, len(nodes)),
+		Adj:   make([][]int, len(nodes)),
+	}
+
+	for key, u := range nodes {
+		graph.Nodes[u] = key
+		graph.Adj[u] = adj[u]
+	}
+
+	return graph, nil
+}
+
+// BetweennessCentrality is a NodeMetric that calculates node betweenness
+// centrality using Brandes' algorithm. Betweenness centrality for each node
+// is the number of shortest paths passing trough that node, not counting
+// shortest paths starting or ending at that node. This is a useful metric
+// to measure control of individual nodes over the whole network.
+type BetweennessCentrality struct {
+	centrality map[NodeID]float64
+}
+
+// NewBetweennessCentralityMetric creates a new BetweennessCentrality instance.
+func NewBetweennessCentralityMetric() *BetweennessCentrality {
+	return &BetweennessCentrality{}
+}
+
+// Name returns the name of the metric.
+func (bc *BetweennessCentrality) Name() string {
+	return "betweeness_centrality"
+}
+
+// betweennessCentrality is the core of Brandes' algorithm. For detailed
+// explanation please read:
+// https://www.cl.cam.ac.uk/teaching/1617/MLRD/handbook/brandes.html
+func betweennessCentrality(g *simpleGraph, node int, centrality []float64) {
```

Comment:
some light comments in this method would help comprehension without needing to read the paper ;)

Reply:
You're totally right, sorry. Changed variable names to make it more readable and added a few comments to help understand what's happening. The final part where we use Brande's MAGIC equation to calculate delta is where it's still hard to understand imho and where reading the "paper" is more important.

Reply:
looks good, thanks!